### ISSUE TITLE
Audit PDOM structure of a screen after construction

### ISSUE THREAD

**Issue by zepumph** [phetsims/joist#511]
From discussion in https://github.com/phetsims/scenery-phet/issues/381 and other places for a11y. It would be really nice to have a place where we can run assertions that individual nodes and the entire screen is constructed for the PDOM as expected. 

For nodes this can't be done as individual options are set because we have to be flexible to option order. For the entire `ScreenView`, it would be nice, once it is constructed, to run basic assertions. For example assert that there are no nodes outside the top level `section` elements (like play area etc. . .)

Tagging @jonathanolson @pixelzoom @samreid @jessegreenberg because they were involved in discussing this.

**Comment by pixelzoom** [comment]
@zepumph What feedback do you want on this for developer meeting?  I agree that we need a way to verify that the PDOM is complete, especially if sim developers are responsible for the PDOM's structure/order.  But I don't have enough familiarity with PhET's a11y implementation to recommend how to verify.

**Comment by zepumph** [comment]
I'm not sure how this made it onto the developers meeting, I think I am just interested in hearing peoples opinions about it, no need to take up dev meeting time for it. For example, there could be a recursive audit function called in `Screen.initializeView()` that cascades to all a11y nodes and makes sure they look as expected on startup. We would also set high level assertions, like "make sure all content is within a section Node."

**Comment by zepumph** [comment]
I implemented an auditing function that is recursively called on all nodes when the screenView is initialized. This function audits only that node, and not its relationship to other nodes, or its structure as it pertains to the whole screen (that will be another step).

I went through all the setters in the mixin and found assertions that made more sense in a context where the Node has been completely constructed. Then added many of these assertions to the audit function.

Here are some more ideas for assertions we can have for each node in this new function:
* You must have either (a) innerContent or (b) children?
* not mixing higher/lower level apis. 

Another question. Many of these asserts are littered throughout the setters, in a pattern like the following for setInputType: "if tagName is defined, make sure it is an input because you can't set inputType on a tagName other than 'input'" or something like that, because we have to be able to maintain any option order. I'm wondering if we should keep those in. It is helpful because the audit function I wrote is only called on construction, so things could dynamically get out of sync. Asserts in the setters can help that, but it seems like neither the audit function or the dispersed assertions in setters are doing the complete work. I think we need both of them, but wanted to mention the duplication (since these assertions have to be in both places, and so maintained in both places).


Good news, adding this auditing didn't throw any errors in fuzzTesting!

@jessegreenberg can you please comment, and potentially add to the list of potential things to validate on construction.

**Comment by zepumph** [comment]
The other piece of this isn't on the individual Node level, but rather on the whole PDOM level. An idea to do this: We could get the HTML as text via `phet.joist.display.accessibleDOMElement.outerHTML`, and then validate different pieces of that. I think this would be easier than trying to parse (a) nodes in a tree, or (b) accessible instances in the tree. 

If we have the HTML, then we can validate that the output is as expected. We could do this on the screen level, or(and) on the sim level. Tagging @jessegreenberg for discussion.

**Comment by zepumph** [comment]
@jessegreenberg had good idea to potentially audit in the animation loop. This would be horrible for performance, so maybe we could do it every N frames, hide it under `assertSlow`.

**Comment by zepumph** [comment]
- [ ] also we should add a blurb at the top of `Accessibility.js` that talks about how it is hard to validate  proper PDOM structure while the Node is being created (to maintain option fleibility), hence the presence of `accessibleAudit`.

**Comment by zepumph** [comment]
More ideas:
* Have a separate file that has a high level schema of what a PhET PDOM should look like, then pass in the accessible DOM element for validation (this is more like HTML validation).
* Test node things about the ScreenView, like make sure that the accessibleOrder is still (1) screenSummary, (2) playArea, (3) control panel.
* Get a library as a preload that will a11y audit the a11y DOM on startup, then strip that out on build. This audit would more be for general good HTML and not PhET specific.

**Comment by zepumph** [comment]
- [ ] There is still a todo for this issue in `ScreenSummaryNode` in regards to this issue.

**Comment by pixelzoom** [comment]
@jessegreenberg reminder that this TODO is still in ScreenSummaryNode.js:

```js
TODO: Add assertions for this, see https://github.com/phetsims/joist/issues/511
```

**Comment by jessegreenberg** [comment]
In this issue, an auditor for PDOM content was added for subtrees of Nodes. In https://github.com/phetsims/perennial/issues/138, work has been done to audit the output PDOM content as a whole.

We considered addressing the duplication of assertions in setters and the `accessibleAudit` function by refactoring them out, but realized that several assertions are not the same in both cases. Instead it may be best to have documentation that references the `accessibleAudit` from the setter function.

So all that is left for this issue is the remaining TODO's and checkboxes listed in the above comments.

**Comment by zepumph** [comment]
Adding another note here. Over in https://github.com/phetsims/sun/issues/444, a design requirement was stated that all AccordionBoxes should have a title in the PDOM, even if their title isn't visible in the visual AccordionBox. This felt to me, at first glance, like a perfect thing to check in `accessibleAudit`. So I overrode accessibleAudit in AccordionBox, like so: 

```diff
Index: js/AccordionBox.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/AccordionBox.js	(revision 48451ef63a725ecc6be9cbf004c35070bad5bbff)
+++ js/AccordionBox.js	(date 1594932356366)
@@ -545,6 +545,17 @@
     this.expandCollapseButton.accessibleName = accessibleName;
   },
 
+  /**
+   * @override
+   * @public
+   */
+  accessibleAudit: function() {
+    Node.prototype.accessibleAudit.call( this );
+    if ( this.accessibleContent && assert ) {
+      assert && assert( this.expandCollapseButton.accessibleName !== null, 'AccordionBox must have a title for the PDOM' );
+    }
+  },
+
   /**
    * Ensures this node is eligible for GC.
    * @public

```

The issue is that this will fail on partially outfitted sims that turn on interactive description by default. For example CCK. Perhaps it would be nice to hide a more strict layer of common code accessibleAudits under a query parameter or package json flag like "validatePDOMStructure". Coincidentally, the PhET-iO team was talking about something almost identical yesterday over in https://github.com/phetsims/tandem/issues/191.

@jessegreenberg, do you see  more cases where it would be nice to verify in code (codify) that common code components follow PDOM design requirements?
